src/util.rs

```diff
@@ -1015,10 +1015,120 @@ pub fn num_cpus() -> usize {
     std::thread::available_parallelism().map_or(1, std::num::NonZeroUsize::get)
 }
 
+/// Strip base directories from absolute paths in preprocessor output.
+///
+/// This function searches for basedir paths in the preprocessor output and
+/// replaces them with relative path markers. When multiple basedirs are provided,
+/// the longest matching prefix is used. This is similar to ccache's CCACHE_BASEDIR.
+///
+/// Path matching is case-insensitive to handle various filesystem behaviors and build system
+/// configurations uniformly across all operating systems. On Windows, this function also handles
+/// paths with mixed forward and backward slashes, which can occur when different build tools
+/// produce preprocessor output.
+///
+/// Only paths that start with one of the basedirs are modified. The paths are expected to be
+/// in the format found in preprocessor output (e.g., `# 1 "/path/to/file"`).
+pub fn strip_basedirs(preprocessor_output: &[u8], basedirs: &[PathBuf]) -> Vec<u8> {
+    if basedirs.is_empty() || preprocessor_output.is_empty() {
+        return preprocessor_output.to_vec();
+    }
+
+    trace!(
+        "Stripping basedirs from preprocessor output with length {}: {:?}",
+        preprocessor_output.len(),
+        basedirs
+    );
+
+    // Find all potential matches for each basedir using fast substring search
+    // Store as (position, length) sorted by position
+    let mut matches: Vec<(usize, usize)> = Vec::new();
+    #[cfg(target_os = "windows")]
+    let preprocessor_output = &normalize_path(preprocessor_output);
+
+    for basedir_path in basedirs.iter() {
+        let basedir_str = basedir_path.to_string_lossy();
+        let basedir = basedir_str
+            .trim_end_matches(|c| c == '/' || c == '\\')
+            .as_bytes();
+        #[cfg(target_os = "windows")]
+        // Case insensitive
+        let basedir = normalize_path(basedir);
+        // Use memchr's fast substring search
+        let finder = memchr::memmem::find_iter(preprocessor_output, &basedir);
+
+        for pos in finder {
+            // Check if this is a valid boundary (start, whitespace, quote, or '<')
+            let is_boundary = pos == 0
+                || preprocessor_output[pos - 1].is_ascii_whitespace()
+                || preprocessor_output[pos - 1] == b'"'
+                || preprocessor_output[pos - 1] == b'<';
```

Comment:
I think this captures `#line` mentions for sure. `__FILE__` is the other concern that I don't see tests for. That can have:

```c
printf("%s\n",__FILE__); // `,` before it
;__FILE__; // `;` before it
{__FILE__} // `{` before it
const char* fname=__FILE__; // `=` before it
log(__FILE__); // `(` before it
```

Also, perhaps do `is_word_boundary(preprocessor_output.get(pos - 1))` to avoid re-indexing all the time. The optimizer *should* be able to factor it out, but it is definitely cleaner to see `prev ==` instead IMO.